sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -1042,6 +1042,47 @@ struct _pi_queue : _pi_object {
 
   // Indicates that the queue is healthy and all operations on it are OK.
   bool Healthy{true};
+
+  // Data structures and methods for optimizing events usage in in-order queues.
+
+  // Copy of the last command event which is suitable for reuse.
+  // It will be put into the cache when new command is submitted to the in-order
+  // queue.
+  pi_event CandidateForReuse = nullptr;
```

Comment:
Can't we be more generic and just have a cache of events to re-use for anything? The in-order discarded events would just get back to the cache much sooner

Reply:
> Can't we be more generic and just have a cache of events to re-use for anything? 
We already have context-level cache for generic reuse but events get there only when they are ready to be destroyed, i.e. not soon enough for in-order queue.

> The in-order discarded events would just get back to the cache much sooner
Per my understanding in this implementation I put discarded events to cache as soon as possible, unfortunately I don't see how they can be reused sooner.
This implementation limits number of used events to 2 per queue (in-order and with discard_events property) if buffers 
 and fallback assert feature is not used in the program.

Reply:
Per our discussion I understand now that you have **more** information for events re-use within the same queue, i.e. you can re-use on the queue since prior commands in the same queue are guaranteeing its completion already (but it is still not known to the SYCL RT, until event is synchronized to the host, that is can be re-used it for a new command on the same queue.

Please add an explanation why we have this "cache" separate from the generic cache of free events in the context.

Reply:
Added comment.